server/lib/nl/common/bad_words.py

```diff
@@ -12,40 +12,138 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import logging
-from typing import Set
+from dataclasses import dataclass
+from dataclasses import field
+from typing import Dict, List
 
 from server.lib.config import GLOBAL_CONFIG_BUCKET
 from shared.lib import gcs
 
 BAD_WORDS_FILE = 'nl_bad_words.txt'
+_DELIM = ':'
+
+
+@dataclass
+class Entry:
+  # Set to true if the key of the Entry is a single word.
+  is_singleton: bool
+
+  # This entry has phrases beginning with the key of Entry.
+  # NOTE: These are full phrases including the word in the key,
+  # for convenience of matching.
+  #
+  # In case of "very bad dog", "very bad cow"
+  # We will have: ["very bad dog", "very bad cow"]
+  #
+  phrases: List[str] = field(default_factory=list)
+
+  # This is a multi-word entry, where we must match every
+  # word (in any order) in the query.  When sorted, the first
+  # word is the key of Entry, and the remaining words are here.
+  #
+  # In case of "idiot:cat:bat", only if idiot, bat and cat
+  # appear in the query (regardless of ordering),
+  # will we match.  We will have: ["cat", "idiot"]
+  other_words: List[str] = field(default_factory=list)
+
+
+@dataclass
+class BadWords:
+  # The key is a word.  It is either a singleton word,
+  # or the first word in a phrase, or the first sorted
+  # word in a multi-word (aka colon-delimited) entry.
+  words: Dict[str, Entry]
 
 
 #
 # Loads a list of bad words from a text file.
 #
-def load_bad_words() -> Set[str]:
+def load_bad_words() -> BadWords:
   local_file = gcs.download_file(bucket=GLOBAL_CONFIG_BUCKET,
                                  filename=BAD_WORDS_FILE)
-  bad_words = set()
+  return load_bad_words_file(local_file)
+
+
+def load_bad_words_file(local_file: str, validate: bool = False) -> BadWords:
+  bad_words = BadWords(words={})
 
   with open(local_file) as fp:
     for line in fp:
-      line = line.strip()
+      line = line.strip().lower()
+      # Remove extra spaces between words
+      line = ' '.join(line.split())
       # Ignore comments
       if line.startswith('#') or line.startswith('//'):
         continue
 
-      bad_words.add(line.lower())
+      # Generally, be resilient to duplicates.  Only assert when `validate` is set.
+      if _DELIM in line:
+        words = sorted([w.strip() for w in line.split(_DELIM) if w.strip()])
+        if words[0] not in bad_words.words:
```

Comment:
what is you have a situation where the same first word is part of multiple DELIM sequences?

Reply:
Great catch!  Added support and a test.